mcorpc/authz_rego_test.go

```diff
@@ -0,0 +1,224 @@
+		})
+
+		Context("When the user agent or caller is right", func() {
+			It("Should succeed", func() {
+				auth, err := authz.authorize()
+				Expect(err).ToNot(HaveOccurred())
+				Expect(auth).To(BeTrue())
+			})
+
+			It("Default policy should fail", func() {
+				authz.agent.meta.Name = "boop"
+				auth, err := authz.authorize()
+
+				Expect(err).ToNot(HaveOccurred())
+				Expect(auth).To(BeFalse())
+			})
+
+		})
+
+		Context("When facts are correct", func() {
+			It("Should succeed", func() {
+
+				authz.agent.meta.Name = "facts"
+				auth, err := authz.authorize()
+				Expect(err).ToNot(HaveOccurred())
+				Expect(auth).To(BeTrue())
+
+			})
+		})
+
+		Context("When classes are present and available", func() {
+			It("Should succeed", func() {
+				authz.agent.meta.Name = "classes"
+				auth, err := authz.authorize()
+
+				Expect(err).ToNot(HaveOccurred())
+				Expect(auth).To(BeTrue())
+			})
+		})
+	})
+
+	Describe("Failing tests", func() {
+		var (
+			defaultFacts = json.RawMessage(`{"stub": false, "buts": true}`)
+		)
+		BeforeEach(func() {
+			is.EXPECT().Facts().Return(json.RawMessage(defaultFacts)).AnyTimes()
+			is.EXPECT().Classes().Return([]string{"charlie", "delta"}).AnyTimes()
+		})
+		Context("When the user agent or caller is wrong", func() {
+			It("Should fail if agent isn't ginkgo", func() {
+				authz.req.CallerID = "not=it"
+				auth, err := authz.authorize()
+
+				Expect(err).ToNot(HaveOccurred())
+				Expect(auth).To(BeFalse())
+			})
+
+			It("Should fail with a default policy", func() {
+				authz.req.CallerID = "not=it"
+				authz.agent.meta.Name = "boop"
+				Expect(authz.agent.Name()).To(Equal("boop"))
+
+				authz.cfg.SetOption("plugin.regopolicy.enable_default", "y")
+				auth, err := authz.authorize()
+
+				Expect(err).ToNot(HaveOccurred())
+				Expect(auth).To(BeFalse())
+			})
+		})
+
+		Context("When the facts don't line up", func() {
+			It("Should fail", func() {
+				authz.agent.meta.Name = "facts"
+				auth, err := authz.authorize()
+
+				Expect(err).ToNot(HaveOccurred())
+				Expect(auth).To(BeFalse())
+			})
+		})
+
+		Context("When classes are not what we expect", func() {
+			It("Should fail", func() {
+				authz.agent.meta.Name = "classes"
+				auth, err := authz.authorize()
+
+				Expect(err).ToNot(HaveOccurred())
+				Expect(auth).To(BeFalse())
+			})
+		})
+	})
+
+	Describe("Agents", func() {
+		var (
+			defaultFacts = json.RawMessage(`{"stub": true, "buts": "big"}`)
+		)
+
+		BeforeEach(func() {
+			is.EXPECT().Facts().Return(json.RawMessage(defaultFacts)).AnyTimes()
+			is.EXPECT().Classes().Return([]string{"alpha", "beta"}).AnyTimes()
+		})
+
+		Context("If agent exists on the server", func() {
+			It("Should succed", func() {
+				authz.agent.meta.Name = "agent"
+				auth, err := authz.authorize()
+
+				Expect(err).ToNot(HaveOccurred())
+				Expect(auth).To(BeTrue())
+			})
+		})
+	})
+
+	/*
+	* I am not exactly sure what it means to be "in provisioning" as far as agents goes
```

Comment:
when in provisioning mode it prepares itself for provisioning via https://github.com/choria-io/provisioning-agent

https://github.com/choria-io/go-choria/blob/5f76408222560c3154561c94c3f12c45de354095/choria/framework.go#L131-L163

it also wont load all the agents in that case only rpcutil, choria_util and provisioning.

one might want to add additional controls over the provisioning mode access but only when provisioning is happening since its so restrictive the rules in that mode wont really apply to any other mode